Enable multi-platform support for operator image builds

ashokpariya0 opened the pull request:
These changes enable building and pushing CNAO operator image for multiple platforms (amd64, s390x, arm64) from a single Dockerfile. Enhanced multi-platform support in the build process by adding a PLATFORMS argument in the Makefile for amd64, s390x, and arm64 architectures. Updated the docker-build-operator target to support builds with docker buildx and podman, and added new targets for creating Docker Buildx builders and pushing multi-platform images.

**What this PR does / why we need it**:
The existing CNAO operator ([link](https://quay.io/repository/kubevirt/cluster-network-addons-operator?tab=tags&tag=latest)) currently supports only the amd64 architecture. These changes extend multi-platform support to include arm64 and s390x architectures, enabling broader compatibility for the CNAO operator image.

**Special notes for your reviewer**:
Please note that Docker and Podman handle multi-architecture image builds differently. As a result, the changes have been made accordingly. Additionally, Go cross-compilation support has been added to build binaries for the target architecture.
Added a PLATFORMS argument in the Makefile to support building images for multiple architectures (amd64, s390x, arm64).
Updated the Dockerfile to dynamically handle multi-architecture builds.
Added Go cross-compilation support in Dockerfile to build binaries for the target architecture.
Enhanced the docker-build-operator target to enable multi-platform builds using both docker buildx and podman.
Introduced a new build-multiarch-operator-image target to build and push multi-platform images using podman.
Added a create-builder target to generate Docker Buildx builders for multi-platform image support.

**Release note**:
```release-note
None
```

oshoval commented on .dockerignore:
```diff
@@ -12,9 +12,9 @@
-pkg
-cmd
-tools
+#pkg
```
leftovers?

ashokpariya0 replied:
No, this change is necessary because we need to build the two binaries listed [here](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/Makefile#L107-L111) inside the Dockerfile in order to support Go cross-compilation for building the target binaries.

oshoval replied:
lets remove the lines then completely ?

ashokpariya0 replied:
Yes, Removed in the latest commit.

oshoval commented on Makefile:
```diff
@@ -13,6 +13,14 @@ OPERATOR_IMAGE ?= cluster-network-addons-operator
 REGISTRY_IMAGE ?= cluster-network-addons-registry
 export OCI_BIN ?= $(shell if podman ps >/dev/null 2>&1; then echo podman; elif docker ps >/dev/null 2>&1; then echo docker; fi)
 TLS_SETTING := $(if $(filter $(OCI_BIN),podman),--tls-verify=false,)
+PLATFORMS ?= linux/amd64,linux/s390x,linux/arm64
```
lets have just amd64 by default, so it wont take time on each regular use ?
the rest can be in an example comment

moreover if we build just one image, we dont need buildx just docker build, much easier to maintain, usually this will be the case, unless on releases and such once time come

ashokpariya0 replied:
@oshoval Thanks for the feedback! I agree that simplifying the default platform to just amd64 will streamline the build process and reduce unnecessary overhead. I'll update the configuration accordingly and move the other platforms into an example comment as you suggested.

As for the use of docker build vs. buildx, that approach makes sense for most cases. I'll switch to docker build for simplicity and reserve buildx for when it's needed.

During releases, we'll set PLATFORMS ?= linux/amd64,linux/s390x,linux/arm64, use buildx, and build and push a multi-platform image for the operator. For all other cases, we'll default to using amd64 as the platform. Does that sound good?

oshoval replied:
sgtm thx

ashokpariya0 replied:
Updated in the latest commit.

oshoval commented on Makefile:
```diff
@@ -13,6 +13,14 @@ OPERATOR_IMAGE ?= cluster-network-addons-operator
+ARCH := $(shell uname -m | sed 's/x86_64/amd64/')
+DOCKER_BUILDER ?= docker-builder
+OPERATOR_IMAGE_TAGGED := $(IMAGE_REGISTRY)/$(OPERATOR_IMAGE):$(IMAGE_TAG)
+null  :=
+space := $(null) #
+comma := ,
```
can we just inline those please?

oshoval commented on Makefile:
```diff
@@ -112,16 +120,30 @@ manifest-templator: $(GO)
+	$(MAKE) build-multiarch-operator-image
+else ifeq ($(OCI_BIN),docker)
+	$(MAKE) create-builder
+	docker buildx build \
+        --build-arg BUILD_ARCH=$(ARCH) \
+        --platform $(PLATFORMS) \
+        -f build/operator/Dockerfile \
+        -t ${OPERATOR_IMAGE_TAGGED} \
+        --push .
```
please consider moving it under create-builder target
unless you think it doesnt fit there?

oshoval commented on Makefile:
```diff
@@ -227,9 +249,33 @@ lint-monitoring:
+
+# Target to build the operator image with support for multiple platforms
+build-multiarch-operator-image:
+	# Remove any existing manifest and image
```
please drop the comments (all over unless it is not trivial)

ashokpariya0 replied:
Done.

oshoval replied:
please remove more comments, there are still some trivial ones
we usually add as few as possible
only for non trivial stuff

oshoval commented on build/operator/Dockerfile:
```diff
@@ -1,20 +1,54 @@
-FROM quay.io/centos/centos:stream9
+# Define build-time arguments
+ARG BUILD_ARCH=amd64            # Default value for architecture is amd64
+ARG GO_VERSION="1.23.3"         # Default Go version is 1.23.3
```
might worth to auto take it from go.mod
there are few network repos that doing something like that

ashokpariya0 replied:
Good suggestion! I'll look into implementing that.

ashokpariya0 replied:
I made the changes according to the script [install-go.sh](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/hack/install-go.sh). With the current setup, we are compiling and building the manager binary using the Go version specified above.

Please note that the script [go-version.sh](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/hack/go-version.sh) returns the Go version in major.minor format. Additionally, the command curl -L -s "https://go.dev/dl/?mode=json" does not list Go 1.21, as it's archived. Before Go 1.22, we would need to download it directly from the archive. However, the install-go.sh script currently downloads the latest Go version and proceeds to compile with it.

I’ve made the same update in the Dockerfile to ensure consistency.

oshoval commented on build/operator/Dockerfile:
```diff
@@ -1,20 +1,54 @@
+
+FROM --platform=linux/${BUILD_ARCH} quay.io/centos/centos:stream9 AS builder
+ARG GO_VERSION
+RUN dnf install -y tar gzip && dnf clean all
+RUN ARCH=$(uname -m | sed 's/x86_64/amd64/') && \
+    curl -L https://go.dev/dl/go${GO_VERSION}.linux-${ARCH}.tar.gz -o go.tar.gz && \
+    tar -C /usr/local -xzf go.tar.gz && \
+    rm go.tar.gz
+
```
maybe using docker ADD can simplify some of this ?

oshoval commented on build/operator/Dockerfile:
```diff
@@ -1,20 +1,54 @@
-COPY build/_output/bin/manager $OPERATOR
-COPY build/_output/bin/manifest-templator $MANIFEST_TEMPLATOR
+
+# Copy the final built binaries into the image
```
please drop comments all over unless non trivial

ashokpariya0 replied:
Done

oshoval commented on hack/build-multiarch-operator.sh:
```diff
@@ -0,0 +1,54 @@
+#!/bin/bash
+
+# Function to check if Docker Buildx is available
```
please drop comments unless non trivial

ashokpariya0 replied:
Done.

oshoval reviewed (commented):
Nice thanks
First pass

oshoval commented on hack/build-multiarch-operator.sh:
might worth to consider doing as close to 
https://github.com/nmstate/kubernetes-nmstate/blob/2e791f9d650fe3cd4844e3964a2fea49688466b9/hack/init-buildx.sh please

ashokpariya0 replied:
Updated in latest commit.
FYI- It would be preferable to pass the builder name at line 30 in the script [init-buildx.sh](https://github.com/nmstate/kubernetes-nmstate/blob/2e791f9d650fe3cd4844e3964a2fea49688466b9/hack/init-buildx.sh#L30).

oshoval replied:
can we make it even closer to that script ?
will be easier to maintain, it still seems too close to the original version if i am not wrong

ashokpariya0 replied:
Yes, I did that.
Just to let you know, I noticed an issue in [this line](https://github.com/nmstate/kubernetes-nmstate/blob/2e791f9d650fe3cd4844e3964a2fea49688466b9/hack/init-buildx.sh#L30) of init-buildx.sh — we are not performing the inspection based on the builder name.

oshoval replied:
thanks, if you think it worth it you can open an issue on nmstate repo
i must work on some other stuff at this very moment, so will take me bit time to review

ashokpariya0 commented:
@oshoval  The job [here](https://github.com/kubevirt/cluster-network-addons-operator/actions/runs/11794285408/job/32851563920?pr=1923) failed with the following error:

make[2]: Leaving directory '/tmp/cnao/cluster-network-addons-operator'
**podman manifest push --tls-verify=false localhost:5000/cluster-network-addons-operator:latest
Error: accepts 2 arg(s), received 1**
make[1]: *** [Makefile:143: docker-push-operator] Error 125
make[1]: Leaving directory '/tmp/cnao/cluster-network-addons-operator'
make: *** [Makefile:168: cluster-operator-push] Error 2
Issue Description:

The error indicates that the podman manifest push command is receiving only one argument (localhost:5000/cluster-network-addons-operator:latest), but it expects two arguments (the manifest and the destination).

The command works as expected in my local environment, where I'm using podman version 4.9.4-rhel.
It seems that the issue arises in the CI pipeline, which may be using a different version of Podman, leading to different behavior. Any thoughts?

oshoval commented:
@phoracek @RamLavi @0xFelix 
please consider also reviewing if you can, to see i dont miss anything

oshoval commented on Makefile:
```diff
@@ -112,16 +124,30 @@ manifest-templator: $(GO)
-	$(OCI_BIN) push ${TLS_SETTING} $(IMAGE_REGISTRY)/$(OPERATOR_IMAGE):$(IMAGE_TAG)
+ifeq ($(OCI_BIN),podman)
+	podman manifest push ${TLS_SETTING} ${OPERATOR_IMAGE_TAGGED} ${OPERATOR_IMAGE_TAGGED}
+else ifeq ($(OCI_BIN),docker)
+ifeq ($(words $(PLATFORM_LIST)), 1)
+	docker push ${TLS_SETTING} ${OPERATOR_IMAGE_TAGGED}
+endif
```
please try to think about a makefile target structure that can be simpler wrt to multi platform and one platform
the logic is split here and on docker-build-operator, it will be better to have it concentrated if possible (i understand it might be not possible because there are stages)

ashokpariya0 replied:
For docker buildx, we need to combine both build and push in a single command because the images are not stored locally by default, unlike with docker build. As a result, we don't need a separate docker push command when working with multi-platform builds. To simplify the logic and maintain consistency, we could consider adding the --push option to docker build as well. What do you think @oshoval

oshoval replied:
might be interesting to try if you think it will simplify it please

ashokpariya0 replied:
Done.

oshoval reviewed (commented):
Thanks,
there are some comments that were not addressed / replied to
please check those as well

0xFelix reviewed (commented):
Perhaps an unpopular opinion, but why do we need to maintain two builds for Podman and Docker? In https://github.com/kubevirt/ssp-operator we went with Podman only.

See https://github.com/kubevirt/ssp-operator/pull/1091

oshoval commented:
> Perhaps an unpopular opinion, but why do we need to maintain two builds for Podman and Docker? In https://github.com/kubevirt/ssp-operator we went with Podman only.
> 
> See [kubevirt/ssp-operator#1091](https://github.com/kubevirt/ssp-operator/pull/1091)

personally i am against dropping docker support
it is convenient it exists, and also on kubevirt we still do that for some jobs
moreoever, cnao depends on lot of other component, so it better to be robust in case something happens and is required

While it is a nice question, not sure if the PR is the right place for it (i.e it shouldn't block it)

0xFelix commented on Makefile:
```diff
@@ -13,6 +13,19 @@ OPERATOR_IMAGE ?= cluster-network-addons-operator
+ARCH := $(shell uname -m | sed 's/x86_64/amd64/')
+DOCKER_BUILDER ?= cnao-docker-builder
+OPERATOR_IMAGE_TAGGED := $(IMAGE_REGISTRY)/$(OPERATOR_IMAGE):$(IMAGE_TAG)
+NULL  :=
+SPACE := $(NULL) #
+COMMA := ,
```
Are these vars really required?

ashokpariya0 replied:
It was needed, but removed it latest commit and moved to shell script.

0xFelix commented on Makefile:
```diff
@@ -112,16 +124,24 @@ manifest-templator: $(GO)
 
 docker-push-operator:
-	$(OCI_BIN) push ${TLS_SETTING} $(IMAGE_REGISTRY)/$(OPERATOR_IMAGE):$(IMAGE_TAG)
+ifeq ($(OCI_BIN),podman)
```
This is called docker-push-operator but uses podman?

0xFelix replied:
@ashokpariya0 Can you comment on this?

ashokpariya0 replied:
Yes, We may need to modify the section at Makefile lines 113-126(https://github.com/kubevirt/cluster-network-addons-operator/blob/main/Makefile#L113-L126), but this can be addressed in a separate PR, as renaming might impact some build processes or jobs. The existing code already supports both Docker and Podman, as indicated in [Makefile line 14](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/Makefile#L14).

0xFelix replied:
With your change this target will only push in case `OCI_BIN` is equal to `podman`, IIUC?

ashokpariya0 replied:
Yes exactly, with docker buildx, push is combined with image build.

0xFelix replied:
Fine for me, @oshoval wdyt?

oshoval replied:
fine by me to address naming on follow-up
thx

ashokpariya0 replied:
@oshoval  I've already looked into this and found that `docker-*` is being called from various scripts in the automation, hack folder, etc. Additionally, we need to update the README and other related files. Including all of this in a single commit would add unnecessary complexity. It would be better to handle these changes separately.

oshoval replied:
ok thanks

ashokpariya0 replied:
@oshoval I will create an separate issue for this, resolve it, and provide a follow-up.

0xFelix commented on Makefile:
```diff
@@ -227,9 +247,44 @@ lint-monitoring:
+build-multiarch-operator-docker:
+ifeq ($(words $(PLATFORM_LIST)), 1)
+	docker build \
+            --build-arg BUILD_ARCH=$(ARCH) \
+            --platform $(PLATFORMS) \
+            -f build/operator/Dockerfile \
+            -t ${OPERATOR_IMAGE_TAGGED} --push .
```
Can you reuse/deduplicate the args?

ashokpariya0 replied:
Updated in latest commit.

0xFelix commented on hack/build-multiarch-operator.sh:
```diff
@@ -0,0 +1,61 @@
+  # If there is no buildx let's install it
+  if ! docker buildx > /dev/null 2>&1; then
+     mkdir -p ~/.docker/cli-plugins
+     curl -L https://github.com/docker/buildx/releases/download/v0.6.3/buildx-v0.6.3.linux-amd64 --output ~/.docker/cli-plugins/docker-buildx
```
Hardcoded buildx version?

ashokpariya0 replied:
good catch, corrected it.

0xFelix commented on hack/build-multiarch-operator.sh:
Did you run shellcheck against this file?

ashokpariya0 replied:
Yes.

0xFelix reviewed (commented):
@oshoval Not to block the PR, but IMO it adds a lot of complexity which I'm not sure is required. (especially looking at `hack/build-multiarch-operator.sh`)

oshoval commented:
> @oshoval Not to block the PR, but IMO it adds a lot of complexity which I'm not sure is required. (especially looking at `hack/build-multiarch-operator.sh`)

we can indeed request to have a separate script for it instead having it internal as it is now, wdyt?

ashokpariya0 commented:
@oshoval @0xFelix The review comments have been addressed. Please let me know if there's anything else I need to do.

0xFelix commented on build/operator/Dockerfile:
```diff
@@ -1,4 +1,25 @@
+
+ARG TARGETOS
+ARG TARGETARCH
+ARG TARGETPLATFORM
```
Is this arg used?

ashokpariya0 replied:
Yes

0xFelix replied:
I may be blind, but where?

ashokpariya0 replied:
we use at https://github.com/kubevirt/cluster-network-addons-operator/blob/106c2b661dff49532a0e2b5738f7c12be41a48f2/build/operator/Dockerfile#L18-L19

ashokpariya0 replied:
TARGETPLATFORM can be removed.

0xFelix commented on hack/build-operator-docker.sh:
```diff
@@ -0,0 +1,23 @@
+    ./hack/init-buildx.sh "$DOCKER_BUILDER"
+    docker buildx build --platform "$PLATFORMS" $BUILD_ARGS
+fi
+
```
Empty lines at EOF

ashokpariya0 replied:
Removed.

0xFelix commented on build/operator/Dockerfile:
```diff
@@ -1,4 +1,25 @@
+WORKDIR /go/src/cluster-network-addons-operator
+COPY . .
+
+ARG TARGETOS
+ARG TARGETARCH
```
Where are these set?

ashokpariya0 replied:
TARGETOS and TARGETARCH are set by docker/podman based on the platform you're building for based on (--platform ) flag.

ashokpariya0 commented:
@oshoval @0xFelix  I’ve addressed all the review comments.
Could you please take another look and proceed with the next steps for the PR?
Let me know if there’s anything else you’d like me to update.

0xFelix reviewed (approved):
/lgtm

oshoval commented on build/operator/Dockerfile:
```diff
@@ -1,4 +1,24 @@
+
+RUN dnf install -y tar gzip jq && dnf clean all
+RUN ARCH=$(uname -m | sed 's/x86_64/amd64/') && \
+    GO_VERSION=$(curl -L -s "https://go.dev/dl/?mode=json" | jq -r '.[0].version') && \
```
i dont think we want unpinned go version?
stuff can break suddenly

unless we are fine with that or if this was the case already, lets see other maintainers idea

ashokpariya0 replied:
This was already the case; this PR follows the same approach.

ashokpariya0 replied:
@oshoval For this as well, I will create a separate issue, resolve it, and provide a follow-up.

oshoval replied:
thanks, if this was already the case, no rush to fix it, its ok

oshoval replied:
Please consider in the follow-up to use ADD if it helps

oshoval commented on hack/build-operator-docker.sh:
```diff
@@ -0,0 +1,20 @@
+    exit 1
+fi
+
+# Split the comma-separated platforms into an array
```
please drop comments
from all files, unless non trivial

oshoval reviewed (commented):
quick pass

oshoval commented:
please see pull-cluster-network-addons-operator-unit-test

ashokpariya0 commented:
> please see pull-cluster-network-addons-operator-unit-test
>
> Sure.

oshoval commented:
Thanks

@qinqon 
any thing you want to add please ?

ashokpariya0 commented:
Could we please have a release once this PR is merged? There were commits made after the last release (v0.96.0) on October 8th. We would like to perform end-to-end testing on this release.
Also, as discussed earlier, please set PLATFORMS=linux/amd64,linux/s390x before pushing the upstream image, as by default it will only publish amd64-compatible images.

oshoval commented:
yes thanks but at this very moment i cant work on that
once one of us have more time

ashokpariya0 commented:
> yes thanks but at this very moment i cant work on that once one of us have more time
> 
> you are welcome if you want to check project-infra and suggest a PR
>
> Okay, thanks! For now, merging the PR once it passes the tests would be helpful.

oshoval commented:
@RamLavi 
if you want to also take a look, before we approve soon

ashokpariya0 commented:
@RamLavi Would like to add anything ??
@oshoval Could you please approve the changes if no further actions are required, so that it can trigger the other tests?

oshoval commented:
Very nice thanks
/approve

ashokpariya0 commented:
@oshoval Thanks for approving the PR! How can we trigger all the tests for it?

oshoval commented:
/test all

thought it auto triggers once lgtm approve

ashokpariya0 commented:
Thanks @oshoval  all tests have passed. There's one pending test job (tide) that requires the LGTM label.

oshoval commented:
Hi @0xFelix 
Can you please review / lgtm once you are good with it ?
Thanks

0xFelix reviewed (approved):
/lgtm

Thanks!

oshoval commented:
Hi
Please check this error before we advance on any PR

https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_cluster-network-addons-operator/1923/pull-cluster-network-addons-operator-unit-test/1860989134017925120#1:build-log.txt%3A10836

also i dont understand how come some lanes seems they didn't run
I will create a vanilla main PR to see CNAO state

oshoval commented:
I think it will be the best to revert this PR and recreate it (due to the CI problems above)
I will also need time to update my own machine to see it works locally

oshoval commented on hack/build-operator-podman.sh:
```diff
@@ -0,0 +1,24 @@
+
+for platform in "${PLATFORM_LIST[@]}"; do
+    podman build \
+        --no-cache \
```
please drop the no-cache on a follow-up

ashokpariya0 replied:
Done.

oshoval replied:
Thanks

oshoval commented on hack/build-operator-podman.sh:
```diff
@@ -0,0 +1,24 @@
+IFS=',' read -r -a PLATFORM_LIST <<< "$PLATFORMS"
+
+# Remove any existing manifest and image
+podman manifest rm "${OPERATOR_IMAGE_TAGGED}" || true
+podman rmi "${OPERATOR_IMAGE_TAGGED}" || true
```
please consider adding redir of stderr to /dev/null so it will clean up the logs from marking error on CI

ashokpariya0 replied:
Added in PR: https://github.com/kubevirt/cluster-network-addons-operator/pull/1928
